molgenis-data-import/src/main/java/org/molgenis/data/export/EmxExportServiceImpl.java

```diff
@@ -0,0 +1,212 @@
+
+@Service
+public class EmxExportServiceImpl implements EmxExportService {
+
+  private final DataService dataService;
+
+  public EmxExportServiceImpl(DataService dataService) {
+    this.dataService = requireNonNull(dataService);
+  }
+
+  @Override
+  @Transactional(readOnly = true)
+  public void download(List<EntityType> entityTypes, List<Package> packages, File downloadFile) {
+    download(entityTypes, packages, downloadFile, Optional.empty());
+  }
+
+  @Override
+  @Transactional(readOnly = true)
+  public void download(
+      List<EntityType> entityTypes,
+      List<Package> packages,
+      File downloadFile,
+      Optional<Progress> progress) {
+    if (!(entityTypes.isEmpty() && packages.isEmpty())) {
+
+      try (ExcelWriter writer = ExcelWriterFactory.create(downloadFile)) {
+        downloadEmx(entityTypes, packages, writer, progress);
+      } catch (IOException e) {
+        throw new EmxExportException(e);
+      }
+    } else {
+      throw new EmptyDownloadRequestException();
+    }
+  }
+
+  private void downloadEmx(
+      List<EntityType> entityTypes,
+      List<Package> packages,
+      ExcelWriter writer,
+      Optional<Progress> progress)
+      throws IOException {
+    Set<Package> packageSet = new LinkedHashSet<>();
+    Set<EntityType> entityTypeSet = new LinkedHashSet<>();
+    resolveMetadata(entityTypes, packages, packageSet, entityTypeSet);
+
+    if (progress.isPresent()) {
+      // Progress per entity plus package sheet + finished message
+      progress.get().setProgressMax(entityTypes.size() + 2);
+    }
+    writePackageSheets(packageSet, writer, progress);
+    writeEntityTypeSheets(entityTypeSet, writer, progress);
+  }
+
+  protected void resolveMetadata(
+      List<EntityType> entityTypes,
+      List<Package> packages,
+      Set<Package> packageSet,
+      Set<EntityType> entityTypeSet) {
+    resolvePackages(packages, packageSet, entityTypeSet);
+    resolveEntityTypes(entityTypes, entityTypeSet);
+  }
+
+  private void writeEntityTypeSheets(
+      Set<EntityType> entityTypes, ExcelWriter writer, Optional<Progress> progressOptional)
+      throws IOException {
+    writeEntityTypes(entityTypes, writer);
+    for (EntityType entityType : entityTypes) {
+      String progressMessage =
+          MessageSourceHolder.getMessageSource()
+              .getMessage(
+                  "emx_export_progress_message",
+                  new Object[] {entityType.getLabel(LocaleContextHolder.getLocale().getLanguage())},
+                  "Downloading: ",
+                  LocaleContextHolder.getLocale());
+      if (progressOptional.isPresent()) {
+        Progress progress = progressOptional.get();
+        progress.status(progressMessage);
+        progress.increment(1);
+      }
+      writeAttributes(entityType.getAllAttributes(), writer);
+      if (!entityType.isAbstract()) {
+        downloadData(entityType, writer);
+      }
+    }
+  }
+
+  private void resolvePackages(
```

Comment:
What is the difference between the package list and package set?

Reply:
packagelist is what is passed to the service, the set is the version with child packages and without duplicates